includes/abilities-api/class-wp-abilities-registry.php

```diff
@@ -159,17 +147,23 @@ public function register( $name, array $properties = array() ): ?WP_Ability {
 			return null;
 		}
 
-		$ability = new WP_Ability(
+		/**
+		 * Filters the class used to instantiate the ability.
+		 *
+		 * This is helpful for creating complex ability types that are not well expressed by array values and callbacks.
+		 *
+		 * @param string $ability_class           The class name to instantiate the ability. Must extend \WP_Ability.
+		 * @param string $name                    The name of the ability being registered.
+		 * @param array<string,mixed> $properties The properties of the ability being registered. See
+		 *                                        `wp_register_ability()` for the expected structure.
+		 *
+		 * @phpstan-param class-string<\WP_Ability> $ability_class
+		 */
+		$ability_class = apply_filters( 'wp_ability_class', WP_Ability::class, $name, $properties );
```

Comment:
What do you think about making it easier use with the following?
```suggestion
		$ability_class = WP_Ability::class;
		if ( isset( $properties['ability_class'] ) } {
				$ability_class = $properties['ability_class'];
				unset( $properties['ability_class'] );
		}
		$ability_class = apply_filters( 'wp_ability_class', $ability_class, $name, $properties );
```

Technically speaking, this could work without a filter, too.

Reply:
Done in https://github.com/WordPress/abilities-api/pull/21/commits/d40a6f5f91c0f5f4a039d96b5e85049c4138e3e5 🚀